File: Graphics/OpenStack.Graphics/Gfx.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Numerics;
using System.Runtime.CompilerServices;
using System.Threading.Tasks;

[assembly: InternalsVisibleTo("OpenStack.Graphics.OpenGL")]

namespace OpenStack.Graphics
{
    /// <summary>
    /// IObjectManager
    /// </summary>
    public interface IObjectManager<Object, Material, Texture>
    {
        Object CreateObject(string path, out object tag);
        void PreloadObject(string path);
    }

    /// <summary>
    /// IModel
    /// </summary>
    public interface IModel
    {
        IDictionary<string, object> Data { get; }
        IVBIB RemapBoneIndices(IVBIB vbib, int meshIndex);
    }

    /// <summary>
    /// IParticleSystem
    /// </summary>
    public interface IParticleSystem
    {
        IDictionary<string, object> Data { get; }
        IEnumerable<IDictionary<string, object>> Renderers { get; }
        IEnumerable<IDictionary<string, object>> Operators { get; }
        IEnumerable<IDictionary<string, object>> Initializers { get; }
        IEnumerable<IDictionary<string, object>> Emitters { get; }
        IEnumerable<string> GetChildParticleNames(bool enabledOnly = false);
    }

    /// <summary>
    /// IShaderManager
    /// </summary>
    public interface IShaderManager<Shader>
    {
        public Shader LoadShader(string path, IDictionary<string, bool> args = null);
        public Shader LoadPlaneShader(string path, IDictionary<string, bool> args = null);
    }

    /// <summary>
    /// ITextureManager
    /// </summary>
    public interface ITextureManager<Texture>
    {
        Texture DefaultTexture { get; }
        Texture BuildSolidTexture(int width, int height, params float[] rgba);
        Texture BuildNormalMap(Texture source, float strength);
        Texture LoadTexture(object key, out object tag, Range? level = null);
        void PreloadTexture(string path);
        void DeleteTexture(object key);
    }

    /// <summary>
    /// IMaterialManager
    /// </summary>
    public interface IMaterialManager<Material, Texture>
    {
        ITextureManager<Texture> TextureManager { get; }
        Material LoadMaterial(object key, out IDictionary<string, object> data);
        void PreloadMaterial(string path);
    }

    ///// <summary>
    ///// MaterialType
    ///// </summary>
    //public enum MaterialType { None, Default, Standard, BumpedDiffuse, Unlit }

    /// <summary>
    /// IMaterial
    /// </summary>
    public interface IMaterial
    {
        string Name { get; }
        string ShaderName { get; set; }
        IDictionary<string, object> Data { get; }
        IDictionary<string, bool> GetShaderArgs();
    }

    /// <summary>
    /// IFixedMaterial
    /// </summary>
    public interface IFixedMaterial : IMaterial
    {
        string MainFilePath { get; }
        string DarkFilePath { get; }
        string DetailFilePath { get; }
        string GlossFilePath { get; }
        string GlowFilePath { get; }
        string BumpFilePath { get; }
        bool AlphaBlended { get; }
        int SrcBlendMode { get; }
        int DstBlendMode { get; }
        bool AlphaTest { get; }
        float AlphaCutoff { get; }
        bool ZWrite { get; }
    }

    /// <summary>
    /// IParamMaterial
    /// </summary>
    public interface IParamMaterial : IMaterial
    {
        Dictionary<string, long> IntParams { get; }
        Dictionary<string, float> FloatParams { get; }
        Dictionary<string, Vector4> VectorParams { get; }
        Dictionary<string, string> TextureParams { get; }
        Dictionary<string, long> IntAttributes { get; }
        //Dictionary<string, float> FloatAttributes { get; }
        //Dictionary<string, Vector4> VectorAttributes { get; }
        //Dictionary<string, string> StringAttributes { get; }
    }

    /// <summary>
    /// IOpenGraphic
    /// </summary>
    public interface IOpenGraphic
    {
        //object Source { get; }
        Task<T> LoadFileObject<T>(string path);
        void PreloadTexture(string texturePath);
        void PreloadObject(string filePath);
    }

    /// <summary>
    /// IOpenGraphicAny
    /// </summary>
    public interface IOpenGraphicAny<Object, Material, Texture, Shader> : IOpenGraphic
    {
        ITextureManager<Texture> TextureManager { get; }
        IMaterialManager<Material, Texture> MaterialManager { get; }
        IObjectManager<Object, Material, Texture> ObjectManager { get; }
        IShaderManager<Shader> ShaderManager { get; }
        Texture LoadTexture(string path, out object tag, Range? rng = null);
        Object CreateObject(string path, out object tag);
        Shader LoadShader(string path, IDictionary<string, bool> args = null);
    }

    /// <summary>
    /// PlatformStats
    /// </summary>
    public static class PlatformStats
    {
        static readonly bool _HighRes = Stopwatch.IsHighResolution;
        static readonly double _HighFrequency = 1000.0 / Stopwatch.Frequency;
        static readonly double _LowFrequency = 1000.0 / TimeSpan.TicksPerSecond;
        static bool _UseHRT = false;

        public static bool UsingHighResolutionTiming => _UseHRT && _HighRes && !Unix;
        public static long TickCount => (long)Ticks;
        public static double Ticks => _UseHRT && _HighRes && !Unix ? Stopwatch.GetTimestamp() * _HighFrequency : DateTime.UtcNow.Ticks * _LowFrequency;

        public static readonly bool Is64Bit = Environment.Is64BitProcess;
        public static bool MultiProcessor { get; private set; }
        public static int ProcessorCount { get; private set; }
        public static bool Unix { get; private set; }

        public static int MaxTextureMaxAnisotropy; // { get; set; }
    }
}

File: Python/openstk/gfx.py
import numpy as np

# typedefs
class IVBIB: pass
class Object: pass
class Material: pass
class Texture: pass
class Shader: pass

# IObjectManager
class IObjectManager:
    def createObject(self, path: str) -> (Object, Object): pass
    def preloadObject(self, path: str) -> None: pass

# IModel
class IModel:
    data: dict[str, object]
    def remapBoneIndices(self, vbib: IVBIB, meshIndex: int) -> IVBIB: pass

# IParticleSystem
class IParticleSystem:
    data: dict[str, object]
    renderers: list[dict[str, object]]
    operators: list[dict[str, object]]
    initializers: list[dict[str, object]]
    emitters: list[dict[str, object]]
    def getChildParticleNames(self, enabledOnly: bool = False) -> list[str]: pass

# IShaderManager
class IShaderManager:
    def loadShader(self, path: str, args: dict[str: bool] = None): pass
    def loadPlaneShader(self, path: str, args: dict[str: bool] = None): pass

# ITextureManager:
class ITextureManager:
    defaultTexture: Texture
    def buildSolidTexture(self, width: int, height: int, rgba: list[float]) -> Texture: pass
    def buildNormalMap(self, source: Texture, strength: float) -> Texture: pass
    def loadTexture(self, key: object, level: range = None) -> (Texture, Object): pass
    def preloadTexture(self, key: object) -> None: pass
    def deleteTexture(self, key: object) -> None: pass

# IMaterialManager
class IMaterialManager:
    textureManager: ITextureManager
    def loadMaterial(self, key: object) -> (Material, Object): pass
    def preloadMaterial(self, key: object) -> None: pass

# IMaterial
class IMaterial:
    name: str
    shaderName: str
    data: dict[str, object]
    def getShaderArgs(self) -> dict[str, bool]: pass

# IFixedMaterial
class IFixedMaterial(IMaterial):
    mainFilePath: str
    darkFilePath: str
    detailFilePath: str
    glossFilePath: str
    glowFilePath: str
    bumpFilePath: str
    alphaBlended: bool
    srcBlendMode: int
    dstBlendMode: int
    alphaTest: bool
    alphaCutoff: float
    zwrite: bool

# IParamMaterial
class IParamMaterial(IMaterial):
    intParams: dict[str, int]
    floatParams: dict[str, float]
    vectorParams: dict[str, np.ndarray]
    textureParams: dict[str, str]
    intAttributes: dict[str, int]

# IOpenGraphic:
class IOpenGraphic:
    def loadFileObject(self, path: str): pass
    def preloadTexture(self, texturePath: str): pass
    def preloadObject(self, filePath: str): pass

# IOpenGraphicAny
class IOpenGraphicAny(IOpenGraphic):
    textureManager: ITextureManager
    materialManager: IMaterialManager
    objectManager: IObjectManager
    shaderManager: IShaderManager
    def loadTexture(self, path: str, level: range = None) -> (Texture, dict[str, object]): pass
    def createObject(self, path: str) -> (Object, dict[str, object]): pass
    def loadShader(self, path: str, args: dict[str, bool] = None) -> Shader: pass

# PlatformStats:
class PlatformStats:
    maxTextureMaxAnisotropy: int = 0
